### PR TITLE
Fix cache issue

### LINKED ISSUE #21: json generator bug if using cache
JSON generation process is not smart enough to realize that it doesn't need to rebuild everything.

It should be as simple as: if you can't find an Environment for a file (or maybe there is a better way to know if it was parsed, like using the `ParseQueue`, which has the list of files that needed to be parsed), then skip - the JSON should still be there and should still be up to date (I suppose we could throw an error if we find that the JSON file is missing for some reason)

### PR DESCRIPTION
**Pull request by nikophil**
fixes #21 & #10

### CODE REVIEW

File: src/BuildContext.php [hunk 1]
```diff
@@ -30,12 +31,13 @@ public function __construct(
         $this->symfonyDocUrl  = $symfonyDocUrl;
     }
 
-    public function initializeRuntimeConfig(string $sourceDir, string $htmlOutputDir, ?string $jsonOutputDir = null, ?string $parseSubPath = null)
+    public function initializeRuntimeConfig(string $sourceDir, string $htmlOutputDir, ?string $jsonOutputDir = null, ?string $parseSubPath = null, ?bool $disableCache = false)
```
Comment (weaverryan) [inline src/BuildContext.php]: `?bool` can just be `bool` i think

File: src/Generator/GeneratorTrait.php [hunk 2]
```diff
@@ -44,4 +81,22 @@ private function getMeta(array $environments, string $parserFilename): MetaEntry
...
+    /**
+     * @param Environment[] $environments
+     */
+    private function useCacheForFile(string $parserFilename): bool
+    {
+        if ($this->buildContext->getDisableCache()) {
+            return false;
+        }
+
+        // if an environment exits for the given file, it means that the file has been built from the parser
```
Comment (weaverryan) [inline src/Generator/GeneratorTrait.php]: exists

And we can:

```php
return !isset($this->environments[$parserFilename]);
```

File: src/Generator/GeneratorTrait.php [hunk 3]
```diff
@@ -44,4 +81,22 @@ private function getMeta(array $environments, string $parserFilename): MetaEntry
 
         return $allMetas[$parserFilename];
     }
+
+    /**
+     * @param Environment[] $environments
+     */
```
Comment (weaverryan) [inline src/Generator/GeneratorTrait.php]: `@param` is out of date

File: src/Generator/JsonGenerator.php [hunk 4]
```diff
@@ -17,42 +18,55 @@ class JsonGenerator
...
-    ) {
-        $environments = $this->extractEnvironments($documents);
+    public function generateJson(array $documents, ProgressBar $progressBar)
+    {
+        $this->extractEnvironmentsAndCachedMetas($documents);
+
+        if (!$this->buildContext->getDisableCache()) {
+            $cachedMetasLoader = new CachedMetasLoader();
+            $cachedMetasLoader->loadCachedMetaEntries($this->buildContext->getHtmlOutputDir(), $metas = new Metas());
+        }
```
Comment (weaverryan) [inline src/Generator/JsonGenerator.php]: Is this needed now? Can't we use `$this->metas` after calling `extractEnvironmentsAndCachedMetas()`?

File: src/Generator/GeneratorTrait.php (hunk 3 above)
Comment (nikophil) [reply src/Generator/GeneratorTrait.php]: even if we expect an array of objects ?
this is usefull for phpstorm to do some completion...

File: src/Generator/JsonGenerator.php (hunk 4 above)
Comment (nikophil) [reply src/Generator/JsonGenerator.php]: yes that's a mistake, sorry

Comment (nikophil) [reply src/Generator/JsonGenerator.php]: fixed

File: src/Generator/GeneratorTrait.php (hunk 2 above)
Comment (nikophil) [reply src/Generator/GeneratorTrait.php]: fixed

File: src/BuildContext.php (hunk 1 above)
Comment (nikophil) [reply src/BuildContext.php]: done

**Pull request merged**